### ISSUE TITLE
Chapter: Promises for asynchronous programming

### ISSUE THREAD

**Issue by rauschma** [rauschma/exploring-js#27]

**Comment by VernonHawk** [comment]
1. Typo in `37.1.4. States of promises`
> - Once a Promise is settled, the settlement value (result or error) is cached. Thus, if .then() or .catch() are called after the settlement, they receive the ~chaed~ **cached** value.

2. A mistake in `37.6.4. Nesting per se is not evil`
> In the following code, we benefit from ~_not_~ nesting:

``` js
db.open()
.then(connection => { // (A)
  return connection.select({ name: 'Jane' })
  .then(result => { // (B)
    // Process result
    // Use `connection` to make more queries
  })
  // ···
  .catch(error => {
    // handle errors
  })
  .finally(() => {
    connection.close(); // (C)
  });
})
```

But this code shows an example of **good** nesting.

**Comment by rauschma** [comment]
@VernonHawk Thanks! Both things will be fixed in the next release.

**Comment by maksverver** [comment]
“37.6.4 Not all nesting is bad” shows how nesting can be beneficial. The code sample uses `finally()`, which hasn't been discussed before, so it's not obvious what the code does.

The `finally()` method should probably be discussed earlier in the chapter.

For completeness sake, it might also make sense to mention the second (optional) parameter to `then()`.

**Comment by Vishwa-code** [comment]
In 37.1.1 It's mentioned,
"A Promise-based function returns a Promise and sends it a result or an error (if and when it is done). The Promise passes it on to the relevant callbacks.",
(But the promise based function 'addAsync()' only returns Promise and it is inside the promise the result or error is evaluated and passed on via resolve or reject, which is handled by call-backs of .then() or .catch().)
The aforementioned statement in quotes is confusing. Can you please make it more clear by being more specific with an example.

**Comment by danzelito** [comment]
40.5.2.3 A simple implementation of Promise.all() 
function all(iterable) {
...
const result = []; // (instead of let result;)

**Comment by smith558** [comment]
The section "43.1 The basics of using Promises" doesn't at all mention that `then` can actually take up to 2 arguments. This is not ideal for a beginner learner.

**Comment by rauschma** [comment]
@smith558 I’ve had good experiences with teaching Promises that way:

* `.catch()` is virtually always preferable to the second `.then()` parameter because the latter doesn’t catch rejections of the first `.then()` parameter and because `.catch()` is more self-descriptive.
* Therefore, I consider the existence of the second `.then()` parameter to be advanced knowledge and wait until the following section to explain it: https://exploringjs.com/js/book/ch_promises.html#thenables-promiselike-objects-advanced

Edit: As a teacher, you often face the following conflict:

1. On one hand, you want learners to know everything.
2. On the other hand, you don’t want to overwhelm them with knowledge they don’t need (at least not initially).

In this case, I decided in favor of #2.

**Comment by smith558** [comment]
> [@smith558](https://github.com/smith558) I’ve had good experiences with teaching Promises that way:
> 
> * `.catch()` is virtually always preferable to the second `.then()` parameter because the latter doesn’t catch rejections of the first `.then()` parameter and because `.catch()` is more self-descriptive.
> * Therefore, I consider the existence of the second `.then()` parameter to be advanced knowledge and wait until the following section to explain it: https://exploringjs.com/js/book/ch_promises.html#thenables-promiselike-objects-advanced

True. It's probably not critical, but a complete omission surprised me.

**Comment by mocanu-razvan** [comment]
There are two minor typos in section [43.1.3 Implementing a Promise-based function](https://exploringjs.com/js/book/ch_promises.html#implementing-promise-based-functions). The last paragraph begins with the sentence "_One we have invoked either these functions, ..._". This should be corrected to "**Once** we have invoked either **of** these functions, ..."

**Comment by arlgnv** [comment]
Should be `Promises`, I guess
<img width="681" height="426" alt="Image" src="https://github.com/user-attachments/assets/b7f27f4f-75d3-4cce-9ab3-b14b1b47fad7" />
